src/os/bluestore/BlueStore.cc

```diff
@@ -16466,11 +16610,141 @@ int BlueStore::_do_alloc_write(
       }
     }
   }
-  ceph_assert(prealloc_pos == prealloc.end());
+  ceph_assert(prealloc_pos == pextents.end());
   ceph_assert(prealloc_left == 0);
   return 0;
 }
 
+void BlueStore::_maybe_reformat_object(Collection* c, OnodeRef& o,
+  uint64_t offset, size_t length, const bufferlist& bl, uint32_t op_flags,
+  const span_stat_t& span_stat)
+{
+  int64_t opt_defragment = 0;
+  int64_t opt_recompress = 0;
+  c->pool_opts.get(pool_opts_t::DEEP_SCRUB_DEFRAGMENT, &opt_defragment);
+  c->pool_opts.get(pool_opts_t::DEEP_SCRUB_RECOMPRESS, &opt_recompress);
+  dout(10) << __func__ << " defragment = " << opt_defragment
+    << " recompress = " << opt_recompress
+    << " span stat {" << span_stat << "}"
+    << dendl;
+  _dump_onode<30>(cct, *o);
+
+  bool might_need_defragment = false;
+  bool might_need_recompress = false;
+  auto need = length;
+  WriteContext wctx;
+
+  if (span_stat.cached == 0 && span_stat.allocated_shared == 0) {
+    if (opt_defragment || opt_recompress) {
+      // will probably need write context
+      _choose_write_options(c, o, op_flags, &wctx);
+    }
+    // do reformat if
+    // - object isn't cached (meaning it's not being written at the moment),
```

Comment:
When we write to object, we keep data in Buffers at least until the data is readable from disk.
This is to make sure that if one reads object he will receive the proper data, leaving BlueStore free to delay storing data to disk (aka deferred writes).
As a result one can always read cached data and it will be the proper data.

Hence - there is no need to exclude objects that have unstored data.

The closest scenario to a problem I could imagine would be:
1) Transaction X to object A submitted
2) Buffers updated, Onode metadata updated
3) Data X is not yet written
4) Deep scrub reformats object A
5) Onode metadata updated (Buffers not changed)
6) Data from reformatting **gets to the disk before** data from transaction
7) Power-off or similar

In this case we would never signal that transaction X is completed, yet it will be there.
But it's not a problem, it is normal.

Could you elaborate what prompted you to exclude objects that are being written to?